お天気APIを活用し、ユーザーが入力した地名の天気を表示させる

TakumiHasegawa21 opened the pull request:
## 実装内容
テキストフィールドに入力した地名で検索をかけ、その地名の天気を表示する機能を追加しました。
掲載情報は以下の通り

- 地名
- 最高気温
- 最低気温 (最高気温と同じ値になっていますが、OpenWeatherAPI側のミスかと思われます)
- 湿度
- 最大風速

## 動画

https://github.com/user-attachments/assets/651aa678-aa22-4054-9b5a-53674663093e





## 今回保留した項目とTODOリスト
- GoogleMapへの導線
- 地名が存在しない場合やネットワークエラー時等のエラー処理
- もう少しUIを凝りたい
- 変数名の整備

## 備考
- 注意点等を記載

TakumiHasegawa21 committed:
初期表示画面をWeatherManagementViewControllerにする

TakumiHasegawa21 committed:
TextFieldの追加

TakumiHasegawa21 committed:
tableViewとCellの紐付け & PINRemoteImageの追加 (混ざってしまった)

TakumiHasegawa21 committed:
セルの高さ指定方法を修正

TakumiHasegawa21 committed:
OpenWeatherAPIのレスポンスを元にエンティティファイルを作成

TakumiHasegawa21 committed:
podsがgit対象に入ってくると面倒くさいのでgitignoreに入れる

TakumiHasegawa21 committed:
OpenWeatherAPIから受け取った情報をとりあえずログに出力

TakumiHasegawa21 committed:
Podsファイル以下をgit対象から外す (ファイル自体は残っている)

TakumiHasegawa21 committed:
APIKit経由でviewModelのloadActionからログ出力させてみる

TakumiHasegawa21 committed:
viewModelで受け取ったAPI情報をセルのラベルに表示させる

TakumiHasegawa21 committed:
ユーザーの入力した地名の天気を表示

TakumiHasegawa21 committed:
セルタップ時のセル色変化を消す

TakumiHasegawa21 committed:
セルのUIを少しポップにしてみる

TakumiHasegawa21 commented on .gitignore:
```diff
@@ -18,44 +14,16 @@ timeline.xctimeline
-#
-# Add this line if you want to avoid checking in source code from the Xcode workspace
-# *.xcworkspace
+Pods/
+*.xcworkspace
```
gitignoreでpodsフォルダ以下はコミットの対象に含まないようにしたんですけど、諸々入ってきていますね
ちょっと煩わしいかと思いますがpodsフォルダ以下の変更は無視していただけると🙏

TakumiHasegawa21 committed:
xcworkspaceまでgitignoreに入れてしまっていたので直す

TakumiHasegawa21 committed:
xcschemesファイルもgit対象外になっていたので追加

TakumiHasegawa21 committed:
コーディングガイドラインに沿ってリファクタ

Ken20215 commented on WeatherInformationTableApp/Sources/Repositories/WeatherRepository.swift:
```diff
@@ -0,0 +1,33 @@
+    // MARK: - ItemRepository Protocol
+    func getWeatherInformation(for city: String) -> Single<WeatherResponse> {
+        let request = NorenAPI.GetWeatherRequest(city: city, apiKey: "a284d183b62eae0d39b4a24d5822c531")
+        return Single<WeatherResponse>.create { observer in
```
APIKeyが直書きになってハードコードになり保守性やセキュリティの観点でリスクがあるので、info.plistなりで定義しRepository に注入する形でお願いしたいです

Ken20215 commented on WeatherInformationTableApp/Sources/ViewModels/WeatherManagementViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+        }
+    
+        // MARK: - Outputs & Actions Elements
+        self.isLoading = loadAction.executing.asDriver(onErrorDriveWith: .empty())
```
`isLoading`が使用されてないみたいです。
こちらロード中はIndicator(くるくる)を表示させるようにしたいです

Ken20215 reviewed (commented):
エラー処理を追加してほしいです。
動画のようにエラー発生した際にアラートを表示させるようにしてほしいです。
スクショはサンプルなので、直感的にエラーしたんだとわかるものであればなんでもいいです。

<img width="300"  src="https://github.com/user-attachments/assets/b993b388-a16c-4ea1-a3bf-4f483112d877" />

Ken20215 commented on WeatherInformationTableApp/Sources/Repositories/WeatherRepository.swift:
```diff
@@ -0,0 +1,33 @@
+final class WeatherRepository: WeatherRepositoryProtocol {
+    // MARK: - ItemRepository Protocol
+    func getWeatherInformation(for city: String) -> Single<WeatherResponse> {
+        let request = NorenAPI.GetWeatherRequest(city: city, apiKey: "a284d183b62eae0d39b4a24d5822c531")
```
APIKeyが直書きになってハードコードになり保守性やセキュリティの観点でリスクがあるので、info.plistなりで定義しRepository に注入する形でお願いしたいです

TakumiHasegawa21 replied:
ちょっとここで苦戦しているのでお待ちください
Info.plist等に移し替えるとapiKeyが参照できなくなっているので調査中です

TakumiHasegawa21 replied:
xcconfigファイルはDebugとRelease両方用意しないといけないというオチでした
xcconfigに記載してInfo.plistから参照させるよう修正しました

Ken20215 replied:
こちらはOKです。

Ken20215 commented on WeatherInformationTableApp/Sources/ViewModels/WeatherManagementViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+        }
+    
+        // MARK: - Outputs & Actions Elements
+        self.isLoading = loadAction.executing.asDriver(onErrorDriveWith: .empty())
```
isLoadingが使用されてないです。
ここはデータ取得する間、`indicator`を使うようにしてほしいです。

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/984a50b612fb8af7cdb1b30cf266a054fcc55d98

Ken20215 commented on WeatherInformationTableApp/Sources/ViewModels/WeatherManagementViewModel.swift:
```diff
@@ -0,0 +1,73 @@
+    
+        // MARK: - Outputs & Actions Elements
+        self.isLoading = loadAction.executing.asDriver(onErrorDriveWith: .empty())
+        self.weather = Property(_weather)
```
weatherをVC側に流すのはDriverにしましょう。
あくまでデータの加工はVM側で行い、VC側はバインドだけに限定するイメージです

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/037a5097aa8e208c9d4fe8cb3e1d3c231428fa17

Ken20215 commented on WeatherInformationTableApp/Sources/ViewControllers/WeatherManagementViewController.swift:
```diff
@@ -6,24 +6,92 @@
-    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-        // Get the new view controller using segue.destination.
-        // Pass the selected object to the new view controller.
+extension WeatherManagementViewController: UITableViewDataSource {
```
ここはRxの拡張機能を使ってCell反映してほしいです。

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/037a5097aa8e208c9d4fe8cb3e1d3c231428fa17

Ken20215 commented on WeatherInformationTableApp/Sources/ViewControllers/WeatherManagementViewController.swift:
```diff
@@ -6,24 +6,92 @@
 //
 
 import UIKit
+import APIKit
```
VC側でAPIKitは使ってないと思うので削除でお願いします

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/29c77b9babd2a5e43095e2db4450a42528361eaf

Ken20215 commented on WeatherInformationTableApp/Sources/Views/WeatherInformationTableViewCell.swift:
```diff
@@ -0,0 +1,45 @@
+        weatherImage.pin_cancelImageDownload()
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
```
`setSelected`を使用してないなら削除でお願いします

TakumiHasegawa21 replied:
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/fa26254a727608fde13666d67b61d693d742b027
こちらで対応済みです

Ken20215 commented on WeatherInformationTableApp/Sources/Views/WeatherInformationTableViewCell.swift:
```diff
@@ -0,0 +1,45 @@
+        minTemperatureLabel.text = "最低気温: \(Int(weatherResponse.main.tempMin))°C"
+        humidityLabel.text = "湿度: \(weatherResponse.main.humidity)%"
+        windSpeedLabel.text = "風速: \(weatherResponse.wind.speed)m/s"
+        if let iconCode = weatherResponse.weather.first?.icon {
```
ここのアイコン画像取り出しはCell側でなくURLだけ受け取りで限定させたいです。
やり方は他にもあるかもしれませんが、iconをVMでアンラップしURLにしてから、単独でCellに渡すようにしましょう。

ここはちょっとCellだけでやっている役割が多いので、URLに変える責務はVMとで分けたい

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/c2455bea23251c3c6647e45374793d1aa3e28c9e

Ken20215 commented on WeatherInformationTableApp/Sources/Views/WeatherInformationTableViewCell.swift:
```diff
@@ -0,0 +1,45 @@
+    // MARK: - Initialize
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        weatherImage.pin_clearImages()
```
pin_cancelImageDownload()＋プレースホルダ で十分みたいです。

<img width="500" src="https://github.com/user-attachments/assets/ff3acaa2-71c4-4889-a89e-b6a523950b1b" />

TakumiHasegawa21 replied:
こちらで対応済みです
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/08db4105e567516ac4d879401bceda9822c5ae76

Ken20215 commented on WeatherInformationTableApp/Sources/API/Noren/Request/OpenWeatherMapRequest.swift:
```diff
@@ -0,0 +1,62 @@
+            let decoder = JSONDecoder()
+            return try decoder.decode(WeatherResponse.self, from: data)
+        }
+    }
```
```
func response(from object: Any, urlResponse: HTTPURLResponse) throws -> Response {
    
        guard let jsonObject = object as? [String: Any] else {
            print("Unexpected object type:", type(of: object))
            throw ResponseError.unexpectedObject(object)
        }

        do {

            let data = try JSONSerialization.data(withJSONObject: jsonObject)

            let decoder = JSONDecoder()

            return try decoder.decode(Response.self, from: data)
        } catch {
            throw ResponseError.unexpectedObject(object)
        }
    }
```

ここはこれでやればシンプルだと思いますが、どうでしょうか？
`do catch`のエラー処理も考慮してます。

OpenWeatherの`/data/2.5/weather`のレスポンスは必ず 辞書（JSON object） で返ってくるみたいです。なので配列チェックは完全に不要なので削除OKです。

TakumiHasegawa21 replied:
こちらで対応済み
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/ffefad89589abafbb64a61938592f242ad6f1836

TakumiHasegawa21 committed:
JSONデコード処理のリファクタ

Ken20215 reviewed (commented):
[Uploading Simulator Screen Recording - iPhone 16 Pro - 2025-09-10 at 10.10.18.mp4…](https://github.com/user-attachments/assets/5099348f-ade9-4aa7-8211-deb280eabbd8)

追加です。地域検索の際にローマ字ならOKでも漢字とかは読み込みは弾かれます。
練習PJではありますが、こちらは修正して欲しいです。

TakumiHasegawa21 committed:
TableViewのデータ連携をRxSwiftで実装

TakumiHasegawa21 committed:
不要なsetSelectedを削除

TakumiHasegawa21 committed:
不要なライブラリ削除

TakumiHasegawa21 committed:
天気画像の取り出しはviewModel側に責務を分散させる

TakumiHasegawa21 committed:
アイコンURLは別ファイルで管理

TakumiHasegawa21 committed:
pin_cancelImageDownload()＋プレースホルダの形式に変更

TakumiHasegawa21 committed:
isLoading実行中はrefreshControlを表示

TakumiHasegawa21 committed:
エラー処理実装

TakumiHasegawa21 committed:
GeocodingAPIを使用し日本語地名の入力にも対応させる

TakumiHasegawa21 committed:
APIキーをxcconfigで管理しセキュリティを担保する

TakumiHasegawa21 committed:
gitignore微修正

TakumiHasegawa21 committed:
Remove sensitive xcconfig files from Git tracking and update .gitignore

- Remove Configurations/Openweather_Secrets.xcconfig from Git tracking
- Remove Configurations/Config-Debug.xcconfig from Git tracking
- Remove Configurations/Config-Release.xcconfig from Git tracking
- Update .gitignore to exclude all xcconfig files except templates
- Keep Openweather_Secrets.xcconfig.template for developer setup

TakumiHasegawa21 committed:
GoogleMapに移動ボタンは使っていないので削除

TakumiHasegawa21 commented:
エラー処理はこちら
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/9b8de8a22adf2d6fbfc0b84bd343ba4e17b4c1fe

漢字対応はこちら
https://github.com/TakumiHasegawa21/WeatherInformationTableApp/pull/2/commits/a9598a61769c895a7d38f9c5329e0b8115433d99

TakumiHasegawa21 committed:
余計なスペースを抹殺

Ken20215 reviewed (commented):
こちら地域によって文字がローマ字表示になったり、エラー表示がされ弾かれてしまいます。
少し確認いただきたいです。

https://github.com/user-attachments/assets/c515b6e8-cc1a-4053-95c8-fdf446fc1458

Ken20215 commented on WeatherInformationTableApp/Sources/ViewModels/WeatherManagementViewModel.swift:
```diff
@@ -0,0 +1,109 @@
+        self.loadAction = Action { city in
+            weatherRepository.getWeatherInformation(for: city).asObservable()
+        }
+
```
現行コードの61行目〜105行目のloadAction.elementsの箇所まで、各コメントを追加して頂き改行をしてください。
ここは開発PJによると思いますが、今回は後の人の読み易さを重視して対応してほしいです。

Ken20215 commented on WeatherInformationTableApp/Sources/ViewModels/WeatherManagementViewModel.swift:
```diff
@@ -0,0 +1,109 @@
+            .bind(to: _city)
+            .disposed(by: disposeBag)
+
+        reload.asObservable()
+            .withLatestFrom(_city) { _, city in city }
+            .flatMap { city in
+                // Geocoding APIで地名を検索（日本語地名も対応）
+                return geocodingRepository.searchCity(city)
+                    .map { responses in
+                        guard let firstResponse = responses.first else {
+                            return city
+                        }
+                        return firstResponse.name
+                    }
+                    .catch { _ in
+                        return Single.just(city)
+                    }
+            }
+            .bind(to: loadAction.inputs)
+            .disposed(by: disposeBag)
```
こちら必ずエラーキャッチされてしまいます。GeocodingRepositoryファイルからデバックしてデータ確認して戻りを追ってみてください🏃‍♂️‍➡️

<img width="1120" height="342" alt="image" src="https://github.com/user-attachments/assets/bd756a92-16e3-47d6-aad3-bb087ccab16e" />
